query-creator/create-query/steps/request.js

```diff
@@ -1,106 +1,59 @@
-            "class_id": 33,
-            "runnable_url": "https://authoring.staging.concord.org/sequences/120"
-          },
-          {
-            "run_remote_endpoint": "https://learn.staging.concord.org/dataservice/external_activity_data/7f7139f9-dced-43f6-93bd-20f5fbf2c642",
-            "class_id": 33,
-            "runnable_url": "https://authoring.staging.concord.org/sequences/120"
-          },
-        ],
-        "user": {
-          "id": "https://example.com/users/1234",
-          "email": "user@example.com"
-        }
-      }),
-      signature: "4d86176489615a0825fbddfead23b6cf4ed75522147ec4e46f0f8d606b7d54d6"
-    }
-  }
+exports.getBody = (event) => {
 
   if (event.body === null) {
     throw new Error("Missing post body in request")
   }
   return queryString.parse(event.body)
 }
 
-exports.validateJSON = (body) => {
+/**
+ * Throws errors if the request body is malformed, and parses the json property in-place.
+ */
+exports.validateRequestBody = (body) => {
+  if (!body.jwt) {
+    throw new Error("Missing JWT");
+  }
   let json = body.json;
   if (!json) {
     throw new Error("Missing json body parameter");
   }
 
-  const signature = body.signature;
-  if (!signature) {
-    throw new Error("Missing signature body parameter");
-  }
-
-  const hmac = crypto.createHmac('sha256', process.env.JWT_HMAC_SECRET);
-  hmac.update(json);
-  const signatureBuffer = new Buffer(signature);
-  const digestBuffer = new Buffer(hmac.digest('hex'));
-
-  if ((signatureBuffer.length !== digestBuffer.length) || !crypto.timingSafeEqual(signatureBuffer, digestBuffer)) {
-    console.log("digestBuffer", digestBuffer.toString())
-    throw new Error("Invalid signature for json parameter");
-  }
-
   try {
     json = JSON.parse(json);
   } catch (e) {
     throw new Error("Unable to parse json parameter");
   }
 
-  if (json.version !== "1.1") {
-    throw new Error(`Request version is ${json.version}, 1.1 required`);
+  if (json.version !== "2") {
+    throw new Error(`Request version is ${json.version}, 2 required`);
   }
 
-  if (!json.learners) {
-    throw new Error("No learners found in json parameter")
-  }
-
-  if (!json.user) {
-    throw new Error("No user info found in json parameter")
-  }
-  if (!json.user.id) {
-    throw new Error("No user id found in json parameter")
-  }
-  if (!json.user.email) {
-    throw new Error("No user email found in json parameter")
-  }
+  body.json = json;
+}
 
-  return json;
+exports.getLearnerDataWithJwt = (reportServiceUrl, queryParams, jwt) => {
+  return axios.post(reportServiceUrl, queryParams,
+    {
+      headers: {
+        "Authorization": `Bearer/JWT ${jwt}`
+      }
+    }
+  ).then((response) => {
+    return response.data;
+  }, (error) => {
+    throw error;
+  });
 }
 
-exports.getRunnables = (json) => {
+/**
+ * Returns [{runnableUrl, learners[]}, ...]
+ */
+exports.getLearnersPerRunnable = (learners) => {
```

Comment:
Maybe I miss understood, but I thought that this in memory loading of learners per runnable wasn't needed anymore?

Reply:
For each batch of learners we receive, we still want to sort them by runnable_url so that we can upload them to the appropriate `queryId` for the appropriate runnable.

That is, unless we don't need this anymore, but I was assuming we're still splitting the query by runnable as we had been doing before.

So the batch of learners does need to be sorted in-memory, but we can forget about them for the next batch.